compiler/optimizer/abstractinterpreter/AbsStackMachineState.cpp

```diff
@@ -0,0 +1,199 @@
+   {
+   TR::AbsOpStack* copy = new (region) TR::AbsOpStack(region);
+   for (size_t i = 0; i < _container.size(); i ++)
+      {
+      copy->_container.push_back(_container[i]->clone(region));
+      }
+   return copy;
+   }
+
+TR::AbsValue* TR::AbsOpStack::pop()
+   {
+   TR_ASSERT_FATAL(size() > 0, "Pop an empty stack!");
+   TR::AbsValue *value = _container.back();
+   _container.pop_back();
+   return value;
+   }
+
+void TR::AbsOpStack::merge(TR::AbsOpStack* other)
+   {
+   TR_ASSERT_FATAL(other->_container.size() == _container.size(), "Stacks have different sizes!");
+
+   for (size_t i = 0; i < _container.size(); i ++)
+      {
+      _container[i]->merge(other->_container[i]);
+      }
+   }
+
+void TR::AbsOpStack::setToTop()
+   {
+   for (size_t i = 0; i <  _container.size(); i ++)
+      {
+      _container[i]->setToTop();
+      }
+   }
+
+void TR::AbsOpStack::print(TR::Compilation* comp)
+   {
+   traceMsg(comp, "Contents of Abstract Operand Stack:\n");
+   
+   const int32_t stackSize = size();
+
+   if (stackSize == 0)
+      {
+      traceMsg(comp, "<empty>\n");
+      traceMsg(comp, "\n");
+      return;
+      }
+   
+   traceMsg(comp, "<top>\n");
+
+   for (int32_t i = 0; i < stackSize; i++) 
+      {
+      TR::AbsValue *value = _container[stackSize - i -1 ];
+      traceMsg(comp, "S[%d] = ", stackSize - i - 1);
+      if (value)
+         value->print(comp);
+      traceMsg(comp, "\n");
+      }
+
+   traceMsg(comp, "<bottom>\n");
+   traceMsg(comp, "\n");
+   }
+
+TR::AbsLocalVarArray* TR::AbsLocalVarArray::clone(TR::Region& region)
+   {
+   TR::AbsLocalVarArray* copy = new (region) TR::AbsLocalVarArray(region);
+   for (size_t i = 0; i < _container.size(); i ++)
+      {
+      copy->_container.push_back(_container[i] ? _container[i]->clone(region) : NULL);
+      }
+   return copy;
+   }
+
+void TR::AbsLocalVarArray::setToTop()
+   {
+   for (size_t i = 0; i < _container.size(); i ++)
+      {
+      if (_container[i])
+         _container[i]->setToTop();
+      }
+   }
+
+void TR::AbsLocalVarArray::merge(TR::AbsLocalVarArray* other)
+   {
+   const int32_t otherSize = other->size();
+   const int32_t selfSize = size();
+   const int32_t mergedSize = std::max(selfSize, otherSize);
+
+   for (int32_t i = 0; i < mergedSize; i++)
+      {
+      TR::AbsValue *selfValue = i < size() ? at(i) : NULL;
+      TR::AbsValue *otherValue = i < other->size() ? other->at(i) : NULL;
+
+      if (!selfValue && !otherValue) 
+         {
+         continue;
+         }
+      else if (selfValue && otherValue) 
+         {
+         TR::AbsValue* mergedVal = selfValue->merge(otherValue);
+         set(i, mergedVal);
+         } 
+      else if (selfValue) 
+         {
+         set(i, selfValue);
+         } 
+      else
+         {
+         set(i, otherValue);
+         }
+      }
+   }
+
+void TR::AbsLocalVarArray::set(uint32_t index, TR::AbsValue *value)
+   {
+   if (size() <= index)
+      {
+      _container.resize(index + 1);
```

Comment:
Yes, NULL values in AbsLocalVarArray are treated as "bottom", which are uninitialized values. I don't know other languages, but for Java, they are just temporarily being NULL. Uninitialized local variables are eliminated by javac, and trying to access them will result in run time error. 

```
int a;
int b = 1;
--------------> [NULL, 1]
a = 3;
--------------->[3, 1]
```

You mentioned this class bring java-specific, moving it down to OpenJ9 I think will be an option

Reply:
So I think removing the ASSERT in `AbsOpStack::push` and let the NULL value get into the stack will be the solution. 

Regarding the `setToTop` for NULL AbsValue, because we may lack the type info, we just create a Top AbsValue with type as TR::NoType

Reply:
`AbsLocalVarArray::setToTop` API is removed. The functionality of `AbsLocalVarArray::setToTop` will be implemented by users (i.e. iterating through the array/stack and then set them to Top. If NULL AbsValues are considered as uninitialized values, inside `AbsLocalVarArray` there isn't a way to set `AbsValue` to Top because `AbsValue` is just an interface